### DISCUSSION TITLE
"invalid argument to '-black-threshold' in maxCircle.bat by snibgo

### DISCUSSION THREAD

**Discussion by ranknfyle** [ImageMagick/ImageMagick#7853] [Help] [answered]
### ImageMagick version

7.1.1-43-Q8-x64

### Operating system, version and so on

Win 10

### Description

Hi, I'm new to IM, but I have programming/img. proc. experience. I came here from the pages of "snibgo", as I want to use his maxCircle.bat script [(](https://im.snibgo.com/maxsqc.htm#maxCircle.bat). In it is a an [fx: expression: 

```
set sMinRad=
if not "%MinRad%"=="" set sMinRad=-black-threshold %%[fx:%MinRad%^+1-epsilon]
```
I assume that %%[fx: etc is some IM expression passed on as an argument. It seems my batch file doesn't resolve this correctly, because I get an error:

```
magick: invalid argument for option '-black-threshold' '-white-threshold' at CLI arg 23 @ error/operation.c/CLISimpleOperatorImage/1870.
```
Note: I had to add a "^" before the "+" since my shell (TCC) needs to escape it. The MinRad arg is set correctly, but sMinRad is not (below). What's the proper syntax of the %%[fx: ... expression, so I can work out any batch-level parsing issues on my end? Thanks.

```
echo MinRad: [%MinRad%]
MinRad: [24]

echo sMinRad [%sMinRad%]
sMinRad [-black-threshold ]
```

**Comment by snibgo** [comment]
The caret is not in the BAT script. The script is for the Windows BAT interpreter. I know nothing about TCC.

Assuming %MinRad% is 24, the characters in the command line that magick should see are:
```
-black-threshold %[fx:24+1-epsilon]
```
Note that when `magick` needs to see a single percent sign `%` the BAT syntax requires us to double it `%%`. Perhaps TCC doesn't need this doubling.

**Comment by ranknfyle** [comment]
Thanks for the reply. I fixed it. TCC is DOS-like shell with many extras. I now get through most of maxCircle.bat. There's one more issue. I want to save locations of the found circles. In the help section on top of maxCircle.bat, I see this:
```
rem %7 IM operations to save the radius map. Default: map is not saved.
@rem    Example operations:
@rem      -write ramap.png
@rem      -auto-level -write ramap.png
@rem      -define quantum:format=floating-point -write ramap.miff
```
But in the actual code, arg %7 seems to mean "saveRam" (save memory, I assume)
```
set SaveRam=%~7
if "%SaveRam%"=="." set SaveRam=
```
Nevertheless, for arg %7 I tried the option below with the result one line lower:
```
-write "[my actual img path here]"

magick: unable to open image '[my actual img path here]': No such file or directory @ error/blob.c/OpenBlob/3596.
```

For arg %7 I also tried the option below with the result one line lower:
```
-define quantum:format=floating-point -write "[my actual miff filepath here]"

magick: no decode delegate for this image format `' @ error/constitute.c/ReadImage/746.
```
Is it possible to get a img/text file with the locations of the circles found? 

BTW, are you available for paid jobs? I have a project that I may be able to deal with myself (I don't know), but I may need help at some point.

**Comment by snibgo** [comment]
"SaveRam" is short for "save the radius map as an image file". The radius map is an image. The text should be one or more valid magick operations. For example, "-write C:\myfiles\abc.png" will create a PNG image file with that name in that directory.

You have:
```
-write "[my actual img path here]"
```
... which would write a file named `[my actual img path here]` but magick can't understand what image format that should be (PNG, JPG, etc). And I suggest that including special characters like [ and ] and space is not a good idea.

>  I want to save locations of the found circles.

The best location for the centre is echoed to the console in:
```
echo Centre at %XC%,%YC% Rad=%Rad%
```
... and written at the end to environment variables. You could write these to a file if you want, eg:
```
echo %XC%,%YC%,%Rad% >myfile.csv
```

> BTW, are you available for paid jobs?

Only if you are in the UK. Otherwise, the admin is too painful.

**Comment by ranknfyle** [comment]
Thanks, I got it to work. maxCircle only finds one max-circle location, right? Is there a way to find all such places?
===========
BTW, here's my explanation of other details. It's just here for the sake of explanation, it's not a question. In my post above, the phrase "[my actual img path here]" was just a placeholder for the purpose of posting here, as I didn't want to post the actual paths for privacy reasons. On my command line, I specified actual, real paths, but I got the errors listed above. Probable cause: since your maxCircle.bat calls other batch files in %PICTBAT%, and on my system "TCC" replaces "cmd", it's likely that these errors are due to un-escaped characters disliked by TCC. Rather than debug this, now I just omit arg %7 and I use "echo %XC%,%YC%,%Rad% >myfile.csv".

**Comment by snibgo** [comment, marked as answer]
Here is an example of a radius map, msc_ex_ram_c.png, taken from [Subimage rectangles](http://im.snibgo.com/subimrect.htm):
![msc_ex_ram_c](https://github.com/user-attachments/assets/ea91c92d-b132-4409-a56e-581a74eb732a)

The pixel at the maximum brightness is at coordinates (846,153). That is the centre of the found circle. Any pixels in the map at less than maximum brightness could have a circle, but that circle would have a smaller radius.

It is possible that many pixels could all have the same maximum brightness.

For example, we make an input image:
```
magick -size 300x100 xc:White -bordercolor Black -border 50 tryrect.png
```
![tryrect](https://github.com/user-attachments/assets/2ad93a67-7e9f-45a2-8dc4-93414dd22928)
Find the maximum circle in the white area:
```
call %PICTBAT%maxCircle tryrect.png abc out.png White . . "-auto-level -write ramap.png"
```
This sets environment variables:
```
set abc_XC=99.5  & set abc_YC=99.5  & set abc_Rad=50
```

out.png is a debugging image that shows the found circle:
![out](https://github.com/user-attachments/assets/5a1088af-eb76-4be1-b4a9-701b7994f576)

We can see that this is one of many possibilities. A circle of the same radius could be drawn further to the right. The radius map, ramap.png, shows this:
![ramap](https://github.com/user-attachments/assets/bfccc285-e166-4cb5-9eb1-fcf36ebff01b)

Where the map image is white, the circle could be drawn. We can list those coordinates by making all non-white pixels transparent, and writing to "sparse-color:" like this:
```
magick ramap.png +transparent White sparse-color:
```
The text output starts like this:
```
99,99,graya(255,1) 100,99,graya(255,1) 101,99,graya(255,1) 102,99,graya(255,1) 103,99,graya(255,1)
```
We should add 0.5 to each of these coordinates. So the circle could be centred at (99.5,99.5) or (100.5,99.5) or (101.5,99.5) etc.

**Comment by ranknfyle** [comment]
Thanks. So "echo %XC%,%YC%,%Rad% >myfile.csv" writes the coords of just one circle. To see all possible locations (what I want), I actually _do_ need to save the radius map "ramap.png" (i.e. use arg %7 above). So, in my specific case, I actually _will_ need to debug the TCC issue further. :) Thanks, all's clear now.

**Comment by snibgo** [comment]
> ... I actually will need to debug the TCC issue further. :)

Debugging the TCC issue is a good idea. However, an alternative is to modify the script. In my maxCircle.bat script, within a magick command, one line says:
```
 %sRAM% ^
```
You could replace that line with:
```
^( +clone -auto-level +transparent White +write sparse-color:abc.txt +delete ^)
```
So, instead of writing a map image file, we write a text file of the coordinates.

This is exactly equivalent to running the script with parameter %7 set to:
```
-auto-level +transparent White +write sparse-color:abc.txt
```

**Comment by ranknfyle** [comment]
Thanks, it works. This will save me lots of debugging :) Is there a way to skip the "srgba(255,255,255,1)" part for each entry, and save the coords only? It's not a problem, I can easily remove the "srgba's" later. But if there's a way to skip it to begin with, I'll use it.

**Comment by snibgo** [comment]
No, I don't think IM can be made to output the coords only.
